lib/std/os/windows.zig

```diff
@@ -898,29 +898,141 @@ pub fn SetFilePointerEx_CURRENT_get(handle: HANDLE) SetFilePointerError!u64 {
 
 pub const GetFinalPathNameByHandleError = error{
     FileNotFound,
-    SystemResources,
     NameTooLong,
     Unexpected,
 };
 
-pub fn GetFinalPathNameByHandleW(
+/// Specifies how to format volume path in the result of `GetFinalPathNameByHandle`.
+/// Defaults to DOS volume names.
+pub const GetFinalPathNameByHandleFormat = struct {
+    volume_name: enum {
+        /// Format as DOS volume name
+        Dos,
+        /// Format as NT volume name
+        Nt,
+    } = .Dos,
+};
+
+/// Returns canonical (normalized) path of handle.
+/// Use `GetFinalPathNameByHandleFormat` to specify whether the path is meant to include
+/// NT or DOS volume name (e.g., `\Device\HarddiskVolume0\foo.txt` versus `C:\foo.txt`).
+/// If DOS volume name format is selected, note that this function does *not* prepend
+/// `\\?\` prefix to the resultant path.
+pub fn GetFinalPathNameByHandle(
     hFile: HANDLE,
-    buf_ptr: [*]u16,
-    buf_len: DWORD,
-    flags: DWORD,
-) GetFinalPathNameByHandleError![:0]u16 {
-    const rc = kernel32.GetFinalPathNameByHandleW(hFile, buf_ptr, buf_len, flags);
-    if (rc == 0) {
-        switch (kernel32.GetLastError()) {
-            .FILE_NOT_FOUND => return error.FileNotFound,
-            .PATH_NOT_FOUND => return error.FileNotFound,
-            .NOT_ENOUGH_MEMORY => return error.SystemResources,
-            .FILENAME_EXCED_RANGE => return error.NameTooLong,
-            .INVALID_PARAMETER => unreachable,
-            else => |err| return unexpectedError(err),
-        }
+    fmt: GetFinalPathNameByHandleFormat,
+    out_buffer: []u16,
+) GetFinalPathNameByHandleError![]u16 {
+    // Get normalized path; doesn't include volume name though.
+    var path_buffer: [PATH_MAX_WIDE * 2]u8 align(@alignOf(FILE_NAME_INFORMATION)) = undefined;
+    var io: IO_STATUS_BLOCK = undefined;
+    var rc = ntdll.NtQueryInformationFile(hFile, &io, &path_buffer, path_buffer.len, FILE_INFORMATION_CLASS.FileNormalizedNameInformation);
```

Comment:
I'm not sure what you mean by turning `FILE_INFORMATION_CLASS` into an enum; it is already [defined as an enum](https://github.com/ziglang/zig/blob/master/lib/std/os/windows/bits.zig#L272). Unless you'd want to be able to return an appropriate struct based on the input `FILE_INFORMATION_CLASS`; e.g., if the user requests `FILE_INFORMATION_CLASS.FileNameInformation`, we return `FILE_NAME_INFORMATION` struct or something. Either way, I'd be happy to work on refactoring this but it'd probably make more sense to do this in a separate PR since this PR didn't really add any more definitions to `FILE_INORMATION_CLASS`. Does that make sense?

Reply:
I've now created a simple wrapper for `ntdll.NtQueryInformationFile`. Changes in cd8e6b6.

Reply:
> I'm not sure what you mean by turning `FILE_INFORMATION_CLASS` into an enum; it is already [defined as an enum](https://github.com/ziglang/zig/blob/master/lib/std/os/windows/bits.zig#L272).

Ah! I had assumed it wasn't because you weren't using enum literal syntax :)